### PR TITLE
Assignments/tic tac toe

### PR DESCRIPTION
**Pull request by EllisWaterman**

### CODE REVIEW

File: assignments/tic-tac-toe/code.js [hunk 1]
```diff
@@ -4,7 +4,111 @@
...
+  if (xs[0][2] == xs[1][1] && xs[0][2] == xs[2][0] && xs[0][2] != null) {
+    drawWinnerLine('d', 1);
+    return xs[0][2];
+  }
+}
+
+
+const winner = () => {
+  let winnersSymbol;
+  if ((winnersSymbol = horizontalWinner()) || (winnersSymbol = verticalWinner()) || (winnersSymbol = diagonalWinner()))
```
Comment (gigamonkey) [inline assignments/tic-tac-toe/code.js:70]: So you're actually relying on the fact that your three winner functions return `undefined` if there's no winner in that direction and the fact that `||` short circuits (i.e. if the left operand  is true it doesn't bother evaluating the right operand. Which is all fine. But given that you can get rid of the variable all together and just write:

```javascript
return horizontalWinner() || verticalWinner() || diagonalWinner();
```

This relies on one fact about `||` that I haven't really taught you which is that it doesn't always evaluate to `true` or `false` but rather to the first "truthy" operand. And `undefined` is "falsey" and any non-empty string is truthy so this works.

File: assignments/tic-tac-toe/code.js [hunk 2]
```diff
@@ -4,7 +4,111 @@
...
+    shape = 'X'
+    colour = 'red'
+  }
+  drawText(shape,
+    col * (width / 3) + xoffset,(row + 1) * (height / 3) - yoffset,colour,size);
+  xs[row][col] = shape;
+  if (winner() != null)
+    endGame(shape + ' WINS');
+  else
+    if (clickCount == 9)
```
Comment (gigamonkey) [inline assignments/tic-tac-toe/code.js:96]: You can bring the `if` up on to the same line as the else and format this like:

```javascript
if (winner !== null) {
  endGame(shape + ' WINS');
} else if (clickCount === 9) {
  endGame('DRAW');
}
```

**Review by gigamonkey — CHANGES_REQUESTED** [review]
Overall a very nice job. I left some comments on things you should fix just so you can see how the code looks after making the changes. But I already raised your grade on booleans and basic control constructs on the basis of this project. Please try the changes I suggested and then request another review.

File: assignments/tic-tac-toe/code.js [hunk 3]
```diff
@@ -4,7 +4,111 @@
...
+let size = Math.min(width, height) * 0.3;
+let xoffset = width / 3 / 2 - (size / 2);
+let yoffset = (height / 3 - 50) / 2;
+let xs = [
+  [null, null, null],
+  [null, null, null],
+  [null, null, null]
+]
+const drawWinnerLine = (lineType, pos) => {
+  if (lineType == 'h') {
```
[the remark this answers is not in the document]
Comment (EllisWaterman) [reply assignments/tic-tac-toe/code.js:22]: fixed

File: assignments/tic-tac-toe/code.js [hunk 4]
```diff
@@ -4,7 +4,111 @@
...
+    let linePosition = height / 3 * pos + (height / 6);
+    drawLine(0, linePosition, width, linePosition, colour, 10);
+  }
+  else if (lineType == 'v') {
+    let linePosition = width / 3 * pos + (width / 6);
+    drawLine(linePosition, 0, linePosition, height, colour, 10);
+  }
+  else
+    if (pos == 0)
```
[the remark this answers is not in the document]
Comment (EllisWaterman) [reply assignments/tic-tac-toe/code.js]: fixed

File: assignments/tic-tac-toe/code.js [hunk 5]
```diff
@@ -4,7 +4,111 @@
...
+let colour;
+let size = Math.min(width, height) * 0.3;
+let xoffset = width / 3 / 2 - (size / 2);
+let yoffset = (height / 3 - 50) / 2;
+let xs = [
+  [null, null, null],
+  [null, null, null],
+  [null, null, null]
+]
+const drawWinnerLine = (lineType, pos) => {
```
[the remark this answers is not in the document]
Comment (EllisWaterman) [reply assignments/tic-tac-toe/code.js]: fixed

File: assignments/tic-tac-toe/code.js (hunk 1 above)
Comment (EllisWaterman) [reply assignments/tic-tac-toe/code.js]: fixed

File: assignments/tic-tac-toe/code.js (hunk 2 above)
Comment (EllisWaterman) [reply assignments/tic-tac-toe/code.js]: fixed

**Review by gigamonkey — APPROVED** [review]
Looks good. Go ahead and merge.

**Pull request merged**